File dialog Enhancements

kenneth-marut-work opened the pull request:
Signed-off-by: Kenneth Marut <kenneth.marut@ericsson.com>

<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.
Contributors guide: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md
-->
![new-file-dialog](https://user-images.githubusercontent.com/62660714/101220372-2e2bc680-364b-11eb-99d4-099becca8629.gif)

#### What it does

Fixes: #2815

Addresses #2815 by improving the browser file dialog to allow for absolute path text input (with *directory* auto suggestion) in addition to the existing select input , also adds a new icon to "Navigate Upward" and fixes minor styling issue with navigation icon focus on enabled/disabled states. Tested on Linux and Windows.

#### How to test
<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->
- Pull this branch and build, start in Browser Mode
- Use the file picker, you should see 2 new icons, an up arrow and a 'edit' icon inside the select input
- Test that the up arrow behaves as it would on a native file picker, also observe that the existing icons' appearance remain disabled even when they are focused (this is not the case on master)
- Use the file picker as you normally would, navigate to various directories
- Switch to the text-input method by clicking the 'edit' icon, you should be brought to the absolute path of where you left off
- Navigate using text input (the input only allows absolute paths), the first matched (alphabetical order) directory will be auto suggested as you type. Press 'Tab' to accept the suggestion and press 'Enter' to navigate to the specified directory, 'Escape' will cancel your input . 
- Observe the backward/forward/up/home behavior, does it behave as expected? 
- Switch back to the select input by blurring the text input, clicking the icon, or by pressing escape. The input should revert to its original state.

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

colin-grant-work commented on packages/filesystem/src/browser/file-dialog/file-dialog-model.ts:
```diff
@@ -88,4 +88,12 @@ export class FileDialogModel extends FileTreeModel {
+
+    canNavigateUpward(): boolean {
+        const treeRoot = this.tree.root;
+        if (FileStatNode.is(treeRoot)) {
```
maybe `return FileStatNode.is(treeRoot) && !treeRoot.uri.path.isRoot` ?

colin-grant-work commented on packages/filesystem/src/browser/style/file-dialog.css:
```diff
@@ -69,12 +72,24 @@
+.dialogContent .theia-NavigationUp.theia-mod-disabled:focus,
+.dialogContent .theia-LocationInputToggle.theia-mod-disabled:focus
+{
+    opacity: var(--theia-mod-disabled-opacity) !important;
```
Why does this need the `!important` annotation? I think it's preferable to avoid that, if it can be avoided.

kenneth-marut-work replied:
Unfortunately this rule is overriding one that is already specified as !important. In this case the icons' opacity is being overridden to retain their disabled appearance even when clicked/focused on

colin-grant-work commented on packages/filesystem/src/browser/style/file-dialog.css:
```diff
@@ -87,17 +102,56 @@
 }
 
+.dialogContent .theia-NavigationUp {
+    left: 61px;
```
There are a lot of very specific pixel values here. Do you see any way to get to a more ergonomic solution (grid, flex) to the placement of these elements?

kenneth-marut-work replied:
I had thought about this too, but in the spirit of how the file dialog currently exists and is styled I decided to retain the absolute positioning as to not introduce too many new changes to the CSS.

colin-grant-work commented:
Very nice work, and a vast improvement over the current dialog. Some thoughts on the functionality:
 - there's an icon to get into editing mode, so I think maybe there should be an icon to get out of editing mode, too.
 - it might be nice to automatically transition to editing mode if the user has focused the select element and then starts typing.
 - I know we discussed the challenges involved, but I think it would be nice to have an auto-complete list / suggestions in edit mode.

kenneth-marut-work commented:
Thanks for taking a look @colin-grant-work. I agree with your requests on keeping the icon around in both input fields, and also adding automatic toggle to input mode when the user starts typing. 
I think after some discussion, autocomplete functionality would be the preferable way to go since the current way this new feature works might not be *too* intuitive (file dialog automatically navigates as path is edited and resolved). I'll spend some exploring this idea.

kenneth-marut-work commented:
> Very nice work, and a vast improvement over the current dialog. Some thoughts on the functionality:
> 
> * there's an icon to get into editing mode, so I think maybe there should be an icon to get out of editing mode, too.
> * it might be nice to automatically transition to editing mode if the user has focused the select element and then starts typing.
> * alternatively, this may be a case for an input-with-select type interaction, where the user can easily use either mode without having to switch explicitly
> * I know we discussed the challenges involved, but I think it would be nice to have an auto-complete list / suggestions in edit mode.

I've pushed a new revision that changes the input behavior to autosuggest the closest matched directory. Additionally I've addressed your comments above:
- Added a icon to toggle back to 'select' mode
- If dropdown is selected and user starts typing, the input field will transition to a the text input

Would you mind taking another look when you have a chance?

colin-grant-work commented on packages/filesystem/src/browser/file-dialog/file-dialog.ts:
```diff
@@ -158,7 +165,7 @@ export abstract class FileDialog<T> extends AbstractDialog<T> {
 
     protected createLocationListRenderer(): LocationListRenderer {
-        return new LocationListRenderer(this.model);
+        return new LocationListRenderer(this.model, undefined, this.fileService);
```
The calls here are a little odd. I think the clarity of the flow would be improved if `createLocationListRenderer()` took an optional `host` argument to be passed into the `new LocationListRenderer()` call, so that it's clear what the `undefined` corresponds to. Then it might make sense just to create the `host` around line 155 rather than waiting for the `LocationListRenderer` to create its own host.

kenneth-marut-work replied:
I like that idea, I've made that change 👌

colin-grant-work commented on packages/filesystem/src/browser/location/location-renderer.tsx:
```diff
@@ -18,34 +18,87 @@ import URI from '@theia/core/lib/common/uri';
             const currentLocation = this.service.location;
             locationList.value = currentLocation ? currentLocation.toString() : '';
+        } else if (locationListTextInput) {
+            setTimeout(() => locationListTextInput.focus());
```
This could use a comment.

kenneth-marut-work replied:
Thanks for pointing this out. I did a bit more experimentation and ended up being able to remove the timeout by adding `preventDefault` on the icon's mouseDown event. What was previously happening was the `locationListTextInput.focus()` would fire before the default behavior of the icon's `onClick`, which was causing the document's `activeElement` to switch from the textInput -> toggleIcon, retriggering a blur event and not properly rendering the textInput. 

I've also slightly refactored the event handlers to make each of their roles a bit more clear.

colin-grant-work commented on packages/filesystem/src/browser/location/location-renderer.tsx:
```diff
@@ -18,34 +18,87 @@ import URI from '@theia/core/lib/common/uri';
+    protected readonly handleTextInputOnBlur = () => this.onTextInputToggle(false);
+    protected readonly handleTextInputMouseDown = (e: React.MouseEvent<HTMLSpanElement>) => this.onTextInputToggle(e.currentTarget.id === 'select-input');
+
     protected doRender(): React.ReactNode {
```
There's a fair bit going on in this method now, so it might benefit from some refactoring. Maybe something like:
```typescript
<>
{!!this.fileService && this.getInputIcons()} // or with the conditional inside the the method
{this.doShowTextInput
  ? this.getTextInput()
  : this.getDropdown()
}
```
That way the high level logic of the conditionals is clear at a glance, and then each subcomponent can be inspected in detail.

kenneth-marut-work replied:
Sounds good, I've made that change

colin-grant-work commented on packages/filesystem/src/browser/location/location-renderer.tsx:
```diff
@@ -104,12 +157,88 @@ export class LocationListRenderer extends ReactRenderer {
+        }
+    }
+
+    protected onLocationKeydown(e: React.KeyboardEvent<HTMLSelectElement>): void {
```
Since the primary effect of this method is to switch to text input from the dropdown, it might do to name it something like `toggleToTextOnKeyDown`. Might there be some way to also count the keypress? At the moment, if you hit a letter, it will transition to text input, but without the letter you pressed.

kenneth-marut-work replied:
I've updated the name of this method. I did think about the 'transition key' value but I'm not sure including the character is *really* necessary. Without a cursor, I don't know what the expected behavior *should* be. For instance, one might expect the entire field to be cleared first and then populated with the 'transition' character vs. the character being appended. Would be happy to discuss further though 😎

colin-grant-work replied:
That's a fair point, and I'm not really sure what the most sensible default is. I'd probably lean toward appending by default, but it might actually push me to retract the suggestion that typing should initiate a transition, since users could be disappointed in whatever default is chosen. Maybe best to just have the user trigger the transition explicitly and then decide what to do with the text input. What do you think?

kenneth-marut-work replied:
I think I'm leaning toward removing the type-to-transition functionality. It's not a very discoverable interaction since the select menu auto-expands when it receives focus, which ends up blocking the custom keyPress behavior. The added functionality might not be worth the complexity.

kenneth-marut-work replied:
I went ahead and removed this functionality

colin-grant-work commented on packages/filesystem/src/browser/location/location-renderer.tsx:
```diff
@@ -18,34 +18,87 @@ import URI from '@theia/core/lib/common/uri';
+        );
     }
 
+    protected onTextInputToggle(shouldShowTextInput: boolean): void {
```
The `on` here might lead one to believe that it's an event rather than an event handler.

kenneth-marut-work replied:
Fair point, I was following the existing convention but agree with what you're saying. I've also updated the other new handlers to have more description method names

colin-grant-work reviewed (commented):
Working very nicely. I have a few minor comments about the code, but I think the functionality is all in place.

kenneth-marut-work commented on packages/filesystem/src/browser/style/file-dialog.css:
```diff
@@ -69,12 +72,24 @@
 .dialogContent .theia-NavigationForward:focus,
-.dialogContent .theia-NavigationHome:focus {
+.dialogContent .theia-NavigationHome:focus,
+.dialogContent .theia-NavigationUp:focus,
```
I believe the original intention of this rule was to remove the focus border for the navigation buttons only. `.theia-NavigationPanel`'s direct children are the nav buttons + the location select menu (which *is* excluded from this rule). With the 2 additional `span` interactables I've added ( `navigateUp`  and `locationInputToggle`) I've applied the same rules. 

Unfortunately, `locationInputToggle` lives inside a wrapper container, so it isn't a direct child of `theia-LocationListPanel`, so a direct-child selector to target all `span` elements wouldn't work. I've adjusted the code as such:

```
.dialogContent .theia-NavigationPanel span:focus 
{
    outline: none;
    border: none;
    box-shadow: none;
}
```
This targets all `span` elements that are children of `.theia-NavigationPanel` (including what is inside the wrapper div). Let me know what you think, I'm happy to go either way.

kenneth-marut-work commented:
Hey @vince-fugnitto would you be able to take a look at this when you have a chance? Thanks!

vince-fugnitto commented:
> Hey @vince-fugnitto would you be able to take a look at this when you have a chance? Thanks!

Sure, I'm back from holiday vacation so I'll add the review to my backlog :)

vince-fugnitto commented on CHANGELOG.md:
```diff
@@ -1,5 +1,7 @@
 # Change Log
 
+- [filesystem] add text input and nagivate up icon to file dialog [#8748](https://github.com/eclipse-theia/theia/pull/8748)
```
nit: it should be placed under a new `## v1.10.0` entry, but I can always update it when preparing the changelog for the upcoming release.

vince-fugnitto commented on packages/filesystem/src/browser/style/file-dialog.css:
```diff
@@ -87,15 +93,52 @@
 .dialogContent .theia-LocationListPanel {
     position: absolute;
-    left: 72px;
+    display: flex;
+    left: 82px;
     top: 1px;
+    width: 417px;
+    height: 21px;
```
Ideally we should not hardcode values (such as height and width) but rather calculate them, or provide variables such as:

https://github.com/eclipse-theia/theia/blob/39887bf9d05c29fedc9466db76ca0a58706a1a2a/packages/core/src/browser/style/menus.css#L17-L26

vince-fugnitto reviewed (commented):
@kenneth-marut-work I have some comments about the code, I'll perform a more in-depth review when they are addressed. I quickly tested the functionality and it worked very well and it is a very welcomed improvement :+1:

vince-fugnitto commented on packages/filesystem/src/browser/file-dialog/file-dialog.ts:
```diff
@@ -116,13 +118,15 @@ export abstract class FileDialog<T> extends AbstractDialog<T> {
     constructor(
         @inject(FileDialogProps) readonly props: FileDialogProps,
-        @inject(FileDialogWidget) readonly widget: FileDialogWidget
+        @inject(FileDialogWidget) readonly widget: FileDialogWidget,
+        @inject(FileService) readonly fileService?: FileService
```
I believe this should be added to the `changelog` as a breaking change since downstream clients who extend the `file-dialog` will be broken. It is also a breaking change due to: https://github.com/eclipse-theia/theia/wiki/Coding-Guidelines#property-injection.

kenneth-marut-work commented:
@vince-fugnitto thanks for the review, I have addressed your comments as follows: 

> nit: it should be placed under a new ## v1.10.0 entry, but I can always update it when preparing the changelog for the upcoming release.

I have updated the changelog with the next release 👍

> Ideally we should not hardcode values (such as height and width) but rather calculate them

I have added variable names for new rules that I have introduced (where acceptable) and also updated some existing rules that were relevant

> I believe this should be added to the changelog as a breaking change since downstream clients who extend the file-dialog will be broken. It is also a breaking change due to: https://github.com/eclipse-theia/theia/wiki/Coding-Guidelines#property-injection.

Thanks for pointing this out, I've made some slight adjustments to the `file-dialog` constructor and the `location-renderer` constructor to make the `FileService` a required constructor argument. I had originally made this optional (along with some conditional logic inside the `location-renderer` to avoid breaking changes (and slightly overcomplicate the logic), but I think it's best to simplify the logic and accept the breaking change. I've pushed this change and updated the changelog.

vince-fugnitto reviewed (commented):
The changes work very well, I tried multiple different use-cases and they all seem correctly handled. If possible I'd like someone else to review as well to see if I may have missed something during the review.

colin-grant-work commented on CHANGELOG.md:
```diff
@@ -1,5 +1,13 @@
+
+- [filesystem] add text input and nagivate up icon to file dialog [#8748](https://github.com/eclipse-theia/theia/pull/8748)
+
+<a name="breaking_changes_1.9.0">[Breaking Changes:](#breaking_changes_1.10.0)</a>
```
Also looks like 'na_giv_ate' -> 'na_vig_ate' would improve l. 5

kenneth-marut-work replied:
@vince-fugnitto thanks for the additional review and catching the typo (also @colin-grant-work). I've pushed the fix 👍

colin-grant-work commented:
Pulled the code to have another look and found that I'm not getting any focus behavior as I tab through the buttons. There's a rule to that effect on master and in the new code:
```
.dialogContent .theia-NavigationPanel span:focus 
{
    outline: none;
    border: none;
    box-shadow: none;
}
```

But I'm not sure that's desirable? Perhaps this rule should only apply, like this one, when those buttons are disabled:

```
.dialogContent .theia-NavigationBack.theia-mod-disabled:focus,
.dialogContent .theia-NavigationForward.theia-mod-disabled:focus,
.dialogContent .theia-NavigationHome.theia-mod-disabled:focus,
.dialogContent .theia-NavigationUp.theia-mod-disabled:focus
{
    opacity: var(--theia-mod-disabled-opacity) !important;
}
```

Otherwise it looks and feels good. Fix the conflict in the CHANGELOG, and it should be good to go.

kenneth-marut-work commented:
> Pulled the code to have another look and found that I'm not getting any focus behavior as I tab through the buttons. There's a rule to that effect on master and in the new code:
> 
> ```
> .dialogContent .theia-NavigationPanel span:focus 
> {
>     outline: none;
>     border: none;
>     box-shadow: none;
> }
> ```
> 
> But I'm not sure that's desirable? Perhaps this rule should only apply, like this one, when those buttons are disabled:
> 
> ```
> .dialogContent .theia-NavigationBack.theia-mod-disabled:focus,
> .dialogContent .theia-NavigationForward.theia-mod-disabled:focus,
> .dialogContent .theia-NavigationHome.theia-mod-disabled:focus,
> .dialogContent .theia-NavigationUp.theia-mod-disabled:focus
> {
>     opacity: var(--theia-mod-disabled-opacity) !important;
> }
> ```
> 
> Otherwise it looks and feels good. Fix the conflict in the CHANGELOG, and it should be good to go.

I've rebased and fixed the changelog, also made some modifications to the CSS to allow the toolbar elements to apply focus only when enabled. See below:
![focusing](https://user-images.githubusercontent.com/62660714/105424057-35fbf600-5c0c-11eb-8944-35d1f64101e5.gif)

The focus is only applied when an *enabled* toolbar item is clicked or tabbed too. Note that the focus border stays visible until something else is focused (even on click). For accessibility reasons, this is nice, but it's not necessarily pretty. 

@vince-fugnitto do you have any input on how to best handle focusing toolbar items in Theia? It seems that in some places focus is applied on click, and not in others. Also note that on master there is no focus behavior for the file dialog toolbar items.

vince-fugnitto commented:
> I've rebased and fixed the changelog, also made some modifications to the CSS to allow the toolbar elements to apply focus only when enabled. See below:
focusing

> The focus is only applied when an enabled toolbar item is clicked or tabbed too. Note that the focus border stays visible until something else is focused (even on click). For accessibility reasons, this is nice, but it's not necessarily pretty.

> @vince-fugnitto do you have any input on how to best handle focusing toolbar items in Theia? It seems that in some places focus is applied on click, and not in others. Also note that on master there is no focus behavior for the file dialog toolbar items.

@kenneth-marut-work @colin-grant-work I'm agree with you that the `outline` is not necessarily pretty but for accessibility when `tabbing` it is likely the best solution, I just wish the icons were all of equal size (square) similarly to vscode:


https://user-images.githubusercontent.com/40359487/105426385-0b646a00-5c19-11eb-9a64-ec0c8edbb4fa.mov

As for `onClick`, I'm thinking we can do something similar to the toolbar items by providing visual feedback ('bounce') to end-users when a toolbar item is clicked (but not the `outline`), similarly to https://github.com/eclipse-theia/theia/pull/6099. Let me know if this is an interesting proposal.


https://user-images.githubusercontent.com/40359487/105426546-672ef300-5c19-11eb-947a-c2b5cbc8f023.mov

kenneth-marut-work commented:
@vince-fugnitto thanks for the suggestions, both are great ideas. I went ahead and pushed a change that includes:
- 'bounce' effect on clicking an active item
- Evenly sized focus borders
- Focus *only* on keyboard access (using the `focus-visible` rule)

![new-behavior](https://user-images.githubusercontent.com/62660714/105523909-9474b300-5ca4-11eb-8116-59d4b0b70b78.gif)

@colin-grant-work and @vince-fugnitto let me know what you think. I know that there is still room to improve the file dialog, but I don't want to go too far beyond the scope of this MR 🙂

vince-fugnitto commented:
> @colin-grant-work and @vince-fugnitto let me know what you think. I know that there is still room to improve the file dialog, but I don't want to go too far beyond the scope of this MR

I think it looks much better, happy the suggestion was useful! :+1: I'll perform another round of review just to be sure that everything works as intended.

vince-fugnitto reviewed (approved):
The changes look good to me :+1:

marechal-p commented on packages/filesystem/src/browser/location/location-renderer.tsx:
```diff
@@ -104,12 +192,77 @@ export class LocationListRenderer extends ReactRenderer {
+            const { value, selectionStart } = e.currentTarget;
+            if (locationTextInput && value.slice(-1) !== '/') {
+                const valueAsURI = new URI(value);
+                const autocompleteDirectories = await this.gatherAlphabetizedDirectories(valueAsURI);
```
My concern here is that we can spin many of those `trySuggestDirectory` async tasks. See a Theia application running with on a high-latency network:

https://user-images.githubusercontent.com/14182238/105744435-1bc36000-5f0b-11eb-9c44-e2c13aabda2b.mp4

In this case, I did `tab` when inputting `projects`, then carried on because nothing happened and when the task finished later it messed up the input. At least it's really odd.

I love the auto-completion idea, but I think we should take care of the concurrency issue here.

marechal-p replied:
I ran the following command if you want to reproduce on localhost:

```sh
sudo tc qdisk add dev lo root netem delay 300ms 10ms 25%
```

I think in this case it was a bit gruesome and you might want to reduce the high value because it takes forever for the frontend to load the way I did it...

kenneth-marut-work replied:
Ah, good catch and definitely not ideal behavior. Thanks for pointing this out. I'll look into this 👌

kenneth-marut-work replied:
Hey @marechal-p, I've implemented a caching system to handle the concurrency issue. The cache is reset whenever the text-input mode is triggered and only will request `fileService.resolve` when a new directory path is detected. The auto suggest will be rendered immediately if a cached value is available, or rendered after a slight delay if a pending promise is resolved. I've made sure to handle the scenario where the text input was being overwritten by a previously resolved promise by ensuring the resolved value is in alignment with the text input.

Thanks for the trick to slow down my machine, here is a screen capture of it under high latency:

![high-latency](https://user-images.githubusercontent.com/62660714/105917469-f52f2300-5ff7-11eb-8476-4721a2062d4a.gif)

And low latency:

![low-latency](https://user-images.githubusercontent.com/62660714/105917593-27d91b80-5ff8-11eb-8b28-a3a62baca1a0.gif)

kenneth-marut-work commented on packages/filesystem/src/browser/location/location-renderer.tsx:
```diff
@@ -18,32 +18,120 @@ import URI from '@theia/core/lib/common/uri';
 
     render(): void {
-        super.render();
+        ReactDOM.render(<React.Fragment>{this.doRender()}</React.Fragment>, this.host, this.doAfterRender);
```
I've added this override to ensure the that the `doAfterRender` will run after a render when the UI state is properly set. For example, ensuring the text input has rendered before applying focus to it. In `ReactWidget`s there is something similar, but it is implemented as a disposable:
https://github.com/eclipse-theia/theia/blob/42042974c59f148974de5e8fd90cba7222cee82f/packages/core/src/browser/widgets/react-widget.tsx#L42

This file is actually a `ReactRenderer` and doesn't contain the same disposable as the `ReactWidget`:
https://github.com/eclipse-theia/theia/blob/42042974c59f148974de5e8fd90cba7222cee82f/packages/core/src/browser/widgets/react-renderer.tsx#L34

But I'm happy to add one in if you think that style is preferable. Thanks!

marechal-p replied:
Thanks for the explanation! Last minor concern is why not just passing `this.doRender()` to `React.render`? Wondering why we need the extra React fragment?

kenneth-marut-work replied:
Ah, just a copy-and-paste from the super call. Not sure why this is the case originally, but TypeScript seems to be happy if I apply the following changes and update the return type from `React.ReactNode` to `React.ReactElement`
![image](https://user-images.githubusercontent.com/62660714/105776312-f7c54600-5f2d-11eb-9829-beb2ac710202.png)
![image](https://user-images.githubusercontent.com/62660714/105776323-001d8100-5f2e-11eb-96ca-64378dd5af2c.png)

I can make that adjustment to `LocationListRenderer` if you'd prefer

marechal-p replied:
Yes please, I always prefer less useless stuff :p

kenneth-marut-work commented on packages/filesystem/src/browser/location/location-renderer.tsx:
```diff
@@ -104,12 +192,77 @@ export class LocationListRenderer extends ReactRenderer {
+        }
+    }
+
+    protected async handleControlKeys(e: React.KeyboardEvent<HTMLInputElement>): Promise<void> {
```
You're right, fixed

colin-grant-work commented on packages/filesystem/src/browser/location/location-renderer.tsx:
```diff
@@ -18,32 +18,191 @@ import URI from '@theia/core/lib/common/uri';
+    tryResolveChildDirectories(inputAsURI: URI): string[] | undefined {
+        const parentDirectory = inputAsURI.path.dir.toString();
+        const input = inputAsURI.path.toString();
+        if (this.cachedDirectories.has(parentDirectory)) {
```
Minor: rather than 
```
if (...has) {
    ...get
}
```
I prefer
```
got = ...get
if (got) {
    use got
}
```
Because doing `has...get` means hashing the same input twice. But can be a little more muddled.

kenneth-marut-work replied:
Good catch, I've cleaned these lines up accordingly. I was also able to remove the last `else if` statement since I'm no longer updating the input value.

colin-grant-work commented on packages/filesystem/src/browser/location/location-renderer.tsx:
```diff
@@ -18,32 +18,191 @@ import URI from '@theia/core/lib/common/uri';
+                const latestInput = this.pendingResolvedDirectories.get(directoryToResolve)?.input || '';
+                const childDirectories = children.filter(child => child.isDirectory)
+                    .map(directory => `${directory.resource.path}/`)
+                    .sort();
```
Do they come out of (alphabetical) order to begin with? Usually `fs.readDir` returns things in alphabetical order, but perhaps `resolve` behaves differently.

kenneth-marut-work replied:
You're right, I've removed `.sort()`

colin-grant-work commented on packages/filesystem/src/browser/location/location-renderer.tsx:
```diff
@@ -18,32 +18,191 @@ import URI from '@theia/core/lib/common/uri';
+        this.toDisposeOnNewCache.dispose();
+        this.directoryCache = new ResolvedDirectoryCache(this.fileService);
+        this.toDisposeOnNewCache.push(this.directoryCache.onDirectoryDidResolve(({ parent, children, input }) => {
+            if (this.locationTextInput?.value === input) {
```
This strategy of input matching seems a bit strange to me. Since `doRenderAutoSuggestion` works as long as the input is a child of the queried directory (and calculates the selection based on the current input, regardless of the `input` field from the `AutoSuggestDataEvent`), why not just check that here, and save the trouble of having to update (or even have)  the `input` field on `AutoSuggestDataEvent`? In fact, `doRenderAutoSuggestion` can mostly run regardless - it may just fail to find a matc, and end up doing nothing.

kenneth-marut-work replied:
Good catch, you're right about the redundant check in `doRenderAutoSuggestion`, which I've renamed to `tryRenderFirstMatch` for clarity. I've changed this a bit to verify that the parent value of the input field matches the parent of the event. This accounts for the scenario under high latency where you might be in the middle of a path and then press backspace several times, for example:

1. start typing:  `/home/kenma`     (auto suggest is pending children of `subdirectory/`)
1. press backspace several times: `/hom` (backspace does not trigger a *new* autosuggest attempt)
1.  autosuggest from step one resolves, but input field's direct parent is now `/` (suggestion will only render if input is a valid match)

Here is a high latency screen cast of the behavior:

![latency](https://user-images.githubusercontent.com/62660714/107567858-dd38d100-6bab-11eb-9b08-ece702b23387.gif)

kenneth-marut-work commented:
Thanks for the comments @colin-grant-work. I've made some changes that address your issues. I've tested in both low and high latency

colin-grant-work commented:
The new changes look good to me 👍

vince-fugnitto reviewed (approved):
I verified the changes and they work really well :+1:
 
I found a future improvement which can be made for both macos and linux. The idea to support `~` as a replacement for the user's home when editing the path (which is something that  is supported in the native dialog).

kenneth-marut-work commented:
> I verified the changes and they work really well 👍
> 
> I found a future improvement which can be made for both macos and linux. The idea to support `~` as a replacement for the user's home when editing the path (which is something that is supported in the native dialog).

Thanks for the review, I'm going ahead with the merge and I'll create an issue regarding the `~` suggestion
